// @intue/core/dist/core/runtime.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Runtime = void 0;
const agent_1 = require("./agent");
/**
 * Runtime environment for INTUE agents
 */
class Runtime {
    /**
     * Create a new runtime instance
     * @param agent - The agent to run
     */
    constructor(agent) {
        this.agent = agent;
    }
    /**
     * Initialize a new INTUE runtime
     * @param config - Runtime configuration
     * @returns Initialized runtime
     */
    static async init(config) {
        const agentConfig = {
            name: config.agentName || 'intue-agent',
            options: config.options
        };
        const agent = new agent_1.Agent(agentConfig);
        return new Runtime(agent);
    }
    /**
     * Start the runtime
     */
    async start() {
        console.log(`Starting INTUE runtime with agent: ${this.agent.getName()}`);
        // Actual startup logic would go here
    }
    /**
     * Stop the runtime
     */
    async stop() {
        console.log('Stopping INTUE runtime');
        // Cleanup logic would go here
    }
    /**
     * Get the agent instance
     */
    getAgent() {
        return this.agent;
    }
}
exports.Runtime = Runtime;
