// @intue/core/dist/modules/sentiment/analyzer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SentimentAnalyzer = void 0;
/**
 * Sentiment analysis module for crypto assets
 */
class SentimentAnalyzer {
    /**
     * Create a new sentiment analyzer
     * @param config - Configuration for the analyzer
     */
    constructor(config = {}) {
        this.providers = config.providers || [];
        this.thresholds = config.thresholds || {
            positive: 0.7,
            negative: 0.3
        };
    }
    /**
     * Add a sentiment provider
     * @param provider - The provider to add
     */
    addProvider(provider) {
        this.providers.push(provider);
    }
    /**
     * Analyze social sentiment for a crypto asset
     * @param symbol - Asset symbol (e.g., 'BTC')
     * @returns Sentiment analysis result
     */
    async analyzeSocialSentiment(symbol) {
        if (this.providers.length === 0) {
            throw new Error('No sentiment providers configured');
        }
        // Collect results from all providers
        const results = await Promise.all(this.providers.map(provider => provider.analyzeSentiment({ symbol })));
        // Average the results
        const averageScore = results.reduce((sum, result) => sum + result.score, 0) / results.length;
        // Determine sentiment classification
        let classification = 'neutral';
        if (averageScore >= this.thresholds.positive) {
            classification = 'positive';
        }
        else if (averageScore <= this.thresholds.negative) {
            classification = 'negative';
        }
        return {
            score: averageScore,
            classification,
            detail: results,
            timestamp: new Date().toISOString()
        };
    }
    /**
     * Analyze market sentiment based on price action
     * @param data - Market data
     * @returns Market sentiment analysis
     */
    analyzeMarketSentiment(data) {
        // This would implement technical analysis for market sentiment
        // For now, we'll return mock data
        return {
            score: 0.65,
            classification: 'bullish',
            indicators: {
                rsi: 68,
                macd: 'positive',
                volume_trend: 'increasing'
            },
            timestamp: new Date().toISOString()
        };
    }
}
exports.SentimentAnalyzer = SentimentAnalyzer;
